README: Improve docs, particularly to help new users (#260)
NordicSemiconductor/zcbor: README.md

Code generation
===============

> Code generation is invoked with the `zcbor code` command:

Reviewer: Should we point out that the test cmake files have examples of these commands?

```sh
zcbor code <--decode or --encode or both> -c <CDDL description file(s)> -t <which CDDL type(s) to expose in the API> --output-cmake <path to place the generated cmake file at>
zcbor code <--decode or --encode or both> -c <CDDL description file(s)> -t <which CDDL type(s) to expose in the API> --oc <path to the generated C file> --oh <path to the generated header file> --oht <path to the generated types header>
```

When you call this, zcbor reads the CDDL files and creates C struct types to match the types described in the CDDL.

> It then creates code that uses the C library to, decode CBOR data into the structs, and/or encode CBOR from the data in the structs.

Reviewer: ```suggestion
It then creates code that uses the C library to decode CBOR data into the structs, and/or encode CBOR from the data in the structs.
```

Finally, it takes the "entry types" (`-t`) and creates a public API function for each of them.
While doing these things, it will make a number of optimizations, e.g. inlining code for small types and removing ununsed functions.

> It outputs the generated code into header and source files and optionally creates a Cmake file to build them.

Reviewer: ```suggestion
It outputs the generated code into header and source files and optionally creates a CMake file to build them.
```

> So, the `zcbor code` command reads one or more CDDL file(s) and generates some or all of these files:

Reviewer: ```suggestion
The `zcbor code` command reads one or more CDDL file(s) and generates some or all of these files:
```

 - A header file with types (always)
 - A header file with declarations for decoding functions (if `--decode`/`-d` is specified)
 - A C file with decoding functions (if `--decode`/`-d` is specified)
 - A header file with declarations for encoding functions (if `--encode`/`-e` is specified)
 - A C file with encoding functions (if `--encode`/`-e` is specified)
 - A Cmake file that creates a library with the generated code and the C library (if `--output-cmake` is specified).

CDDL allows placing restrictions on the members of your data.
Restrictions can be on type (int/string/list/bool etc.), on content (e.g. values/sizes of ints or strings), and repetition (e.g. the number of members in a list).
The generated code will validate the input, which means that it will check all the restriction set in the CDDL description, and fail if a restriction is broken.

There are tests for the code generation in [tests/decode](tests/decode) and [tests/encode](tests/encode).
The tests require [Zephyr](https://github.com/zephyrproject-rtos/zephyr) (if your system is set up to build Zephyr samples, the tests should also build).

Build system
------------

> When calling zcbor with the argument `--output-cmake <file path>`, a Cmake file will be created at that location.

Reviewer: ```suggestion
When calling zcbor with the argument `--output-cmake <file path>`, a CMake file will be created at that location.
```
The proper name is CMake, I guess we should stick with that? Several places in this doc.

> The Cmake file creates a Cmake target library and adds the generated and non-generated source files to it, and adds the relevant include directories.

Reviewer: ```suggestion
The generated CMake file creates a target library and adds the generated and non-generated source files as well as required include directories to it.
```
